internal-dns/types/src/v2/config.rs

```diff
@@ -0,0 +1,249 @@
+use std::collections::HashMap;
+use std::net::{Ipv4Addr, Ipv6Addr};
+
+#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
+pub struct DnsConfigParams {
+    pub generation: Generation,
+    pub time_created: chrono::DateTime<chrono::Utc>,
+    pub zones: Vec<DnsConfigZone>,
+}
+
+impl DnsConfigParams {
+    /// Given a high-level DNS configuration, return a reference to its sole
+    /// DNS zone.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if there are 0 or more than one zones in this
+    /// configuration.
+    pub fn sole_zone(&self) -> Result<&DnsConfigZone, anyhow::Error> {
+        ensure!(
+            self.zones.len() == 1,
+            "expected exactly one DNS zone, but found {}",
+            self.zones.len()
+        );
+        Ok(&self.zones[0])
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema)]
+pub struct DnsConfig {
+    pub generation: Generation,
+    pub time_created: chrono::DateTime<chrono::Utc>,
+    pub time_applied: chrono::DateTime<chrono::Utc>,
+    pub zones: Vec<DnsConfigZone>,
+}
+
+impl DnsConfig {
+    /// Perform a *lossy* conversion from the V2 [`DnsConfig`] to the V1
+    /// [`v1::config::DnsConfig`].  In particular, V2 adds NS and SOA records,
+    /// which did not exist in V1, so they are silently discarded when
+    /// converting down.
+    ///
+    /// If this conversion would leave an empty zone, the zone is omitted
+    /// entirely.
+    pub fn as_v1(self) -> v1::config::DnsConfig {
+        let DnsConfig { generation, time_created, time_applied, zones } = self;
+
+        v1::config::DnsConfig {
+            generation,
+            time_created,
+            time_applied,
+            zones: zones
+                .into_iter()
+                .filter_map(|zone| {
+                    let converted_zone = zone.as_v1();
+                    if converted_zone.records.is_empty() {
+                        None
+                    } else {
+                        Some(converted_zone)
+                    }
+                })
+                .collect(),
+        }
+    }
+}
+
+/// Configuration for a specific DNS zone, as opposed to illumos zones in which
+/// the services described by these records run.
+///
+/// The name `@` is special: it describes records that should be provided for
+/// queries about `zone_name`. This is used in favor of the empty string as `@`
+/// is the name used for this purpose in zone files for most DNS configurations.
+/// It also avoids potentially-confusing debug output from naively printing out
+/// records and their names - if you've seen an `@` record and tools are unclear
+/// about what that means, hopefully you've arrived here!
+#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema, PartialEq, Eq)]
+pub struct DnsConfigZone {
+    pub zone_name: String,
+    pub records: HashMap<String, Vec<DnsRecord>>,
+}
+
+impl DnsConfigZone {
+    fn as_v1(self) -> v1::config::DnsConfigZone {
```

Comment:
I think we discussed elsewhere... I do wonder if we should fail this request if we have any NS records.  We definitely don't want somebody to fetch this, make some other change, then PUT it back without the NS records and not realize they deleted them.

Reply:
yeah, in talking about it, i've come to agree. if there's an NS record in the configuration this returns a [400 with more descriptive message](https://github.com/oxidecomputer/omicron/pull/8047/files#diff-5148dd7859ef6ecea9f31b5c2ecdcb4805f0c062369c9199ee99115c2d4b5448R48-R53) now.